sklearn/cluster/_agglomerative.py

```diff
@@ -453,8 +453,14 @@ def linkage_tree(X, connectivity=None, n_clusters=None, linkage='complete',
         if affinity == 'precomputed':
             # for the linkage function of hierarchy to work on precomputed
             # data, provide as first argument an ndarray of the shape returned
-            # by pdist: it is a flat array containing the upper triangular of
-            # the distance matrix.
+            # by sklearn.metrics.pairwise_distances: it is a flat array
+            # containing the upper triangular of the distance matrix.
+            if len(X.shape) != 2 or X.shape[0] != X.shape[1]:
+                raise ValueError(
+                    'Matrix should be square, '
```

Comment:
I am a bit confused here. I have to say I am not sure what shape `X` should have when `affinity == 'precomputed'`.

The thing is at the moment, the error message says that `X` should be square and the comment above says that `X` should be flat ...

Reply:
@jeremiedbb thinks that the matrix should be square because of the code below:

```py
i, j = np.triu_indices(X.shape[0], k=1)
X = X[i, j]  # now X is a flat array containing the upper triangular of the distance matrix 
```

So the comment needs to be changed.

Reply:
Rather than 
```
provide as first argument an ndarray of the shape returned
by pdist: it is a flat array containing the upper triangular of
by sklearn.metrics.pairwise_distances: it is a flat array
```

say something like:
```
X should be a distance matrix as returned by sklearn.metrics.pairwise_distances 	
```

Reply:
I removed the comment. The input matrix should be squared. And for sure it shouldn't be flat, as there is another check somewhere else asking the user to reshape the potential 1D input as a 2D matrix (n,1).